src/ORM/Table.php

```diff
@@ -1205,36 +1205,88 @@ public function get($primaryKey, $options = [])
+     *   transaction (default: true)
+     * - defaults: Whether to use the search criteria as default values for the new entity (default: true)
+     *
+     * @param array|callable $search The criteria to find an existing record by, or a callable that will
+     *   customize the find query.
      * @param callable|null $callback A callback that will be invoked for newly
      *   created entities. This callback will be called *before* the entity
      *   is persisted.
-     * @return \Cake\Datasource\EntityInterface An entity.
+     * @param array $options The options to use when saving.
+     * @return EntityInterface An entity.
      */
-    public function findOrCreate($search, callable $callback = null)
+    public function findOrCreate($search, callable $callback = null, $options = [])
     {
-        return $this->connection()->transactional(function () use ($search, $callback) {
-            $query = $this->find()->where($search);
-            $row = $query->first();
-            if ($row) {
-                return $row;
-            }
-            $entity = $this->newEntity();
+        $options = array_merge([
+            'atomic' => true,
+            'defaults' => true
+        ], $options);
+
+        if ($options['atomic']) {
+            return $this->connection()->transactional(function () use ($search, $callback, $options) {
+                return $this->_processFindOrCreate($search, $callback, $options);
+            });
+        } else {
+            return $this->_processFindOrCreate($search, $callback, $options);
+        }
+    }
+
+    /**
+     * Performs the actual find and/or create of an entity based on the passed options.
+     *
+     * @param array|callable $search The criteria to find an existing record by, or a callable tha will
+     *   customize the find query.
+     * @param callable|null $callback A callback that will be invoked for newly
+     *   created entities. This callback will be called *before* the entity
+     *   is persisted.
+     * @param array $options The options to use when saving.
+     * @return EntityInterface An entity.
+     */
+    protected function _processFindOrCreate($search, callable $callback = null, $options = [])
+    {
+        $query = $this->find();
+        if (is_callable($search)) {
+            call_user_func($search, $query);
+        } elseif (is_array($search)) {
+            $query->where($search);
+        } else {
+            throw new InvalidArgumentException('Search criteria must be an array or callable');
+        }
+        $row = $query->first();
+        if ($row) {
+            return $row;
+        }
+        $entity = $this->newEntity();
+        if ($options['defaults'] && is_array($search)) {
             $entity->set($search, ['guard' => false]);
-            if ($callback) {
-                $callback($entity);
-            }
-            return $this->save($entity) ?: $entity;
-        });
+        }
+        if ($callback) {
```

Comment:
While we're on the subject of callables. I don't like how it's cake uses the callable parameter hint. It has a more restrictive use case. For example, you can not do this `findOrCreate([$this,'myFind'],[$this,'mySave'])`. Since the parameter hint will complain that it's an array.

I use `call_user_func` in my projects as it lets me use object refs whereas callable forces you to use closures all the time.